### ISSUE TITLE
[inductor] [cpu] `torch.nn.RReLU()` outputs different resutls with eager on cpp backend

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#147255]
### 🐛 Describe the bug

**symptom**: ablation says `aot_eager_decomp_partition` is the first backend with incorrect optimization
**codegen backend**: only CPP backend.

**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.rrelu = torch.nn.RReLU()

    def forward(self, x):
        x = self.rrelu(x)
        return x


model = Model()

x = torch.randn(1, 1, 100, 100)

inputs = [x]


def run_test(model, inputs, device, backend):
    torch.manual_seed(0)
    model = model.to(device)
    inputs = [x.to(device) for x in inputs]
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


device = 'cpu'


output = run_test(model, inputs, device, 'eager')
c_output = run_test(model, inputs, device, 'aot_eager_decomp_partition')
fp64 = run_test(model.to(torch.float64), [x.to(torch.float64) for x in inputs], device, "eager")


print(torch.allclose(output, c_output, 1e-3, 1e-3))
print(torch.max(torch.abs(output - c_output)))
print(torch._dynamo.utils.same(output, c_output, fp64))
```

### Error logs
CPP
```
False
tensor(0.8208)
E0429 17:53:31.189000 6900 site-packages/torch/_dynamo/utils.py:2939] RMSE (res-fp64): 0.06023, (ref-fp64): 0.00000 and shape=torch.Size([1, 1, 100, 100]). res.dtype: torch.float32, multiplier: 2.000000, tol: 0.000100, use_larger_multiplier_for_smaller_tensor: 0
False
```

Triton
```
True
tensor(1.1921e-07, device='cuda:0')
True
```

### Versions
nightly 20250418

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @aakhundov

**Comment by ngimel** [comment]
RRelu is using random numbers to generate output, and compiled and eager functions are not guaranteed to preserve exact randomness pattern, this is expected behavior.

**Comment by shaoyuyoung** [comment]
Actually, I have used `config.fallback_random = True` and `torch.manual_seed(0)` to ensure the RNG op's results.
Feel free to correct me if I am wrong

**Comment by ngimel** [comment]
It looks like rrelu is not registered in lowerings to fallback when random_fallback is set, cc @eellison

**Comment by eellison** [comment]
I believe this is because it is mutating so we need a decomposition for it. or would need native support added to functional tensor cc @bdhirsh.

**Comment by shaoyuyoung** [comment]
I have updated this issue and found that it is a CPP-specific issue.
cc @leslie-fang-intel , mind taking a look at this and assigning it to the right developer?
thanks! :)

**Comment by leslie-fang-intel** [comment]
Hi @CaoE, Could you help to take a look?

**Comment by CaoE** [comment]
@shaoyuyoung There are 2 reasons CPP backend generates different values:
1) Eager always uses double data type for uniform.
https://github.com/pytorch/pytorch/blob/8153340d1073c190ecf221e51abbd983286782a8/aten/src/ATen/native/Activation.cpp#L594-L602 Inductor will decompose rrelu to uniform + mul + where and use the same data type as input https://github.com/pytorch/pytorch/blob/8153340d1073c190ecf221e51abbd983286782a8/aten/src/ATen/native/cpu/DistributionTemplates.h#L272
2) rrelu generates different number of random value on eager and inductor. On eager, rrelu uses one loop to generate random value and the output value if input data <=0. On inductor, uniform will generate same number of random values as input no matter whether input data <=0 or >0, then use mul and where to calculate output.

This gap is due to the decomposition and different implementations on eager and Inductor CPP backend.

Generated code from Inductor CPP backend: 
```

cpp_fused_rrelu_with_noise_functional_0 = async_compile.cpp_pybinding(['float*', 'const float*'], '''
#include "/tmp/torchinductor_ecao/do/cdoggdcp7ux2jv5ebkajvacaprabp6b4h4m2o3zifjj6xwp2kz4n.h"
extern "C"  void kernel(float* in_out_ptr0,
                       const float* in_ptr0)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(10000L); x0+=static_cast<int64_t>(16L))
        {
            {
                if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(10000L)))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                    auto tmp4 = at::vec::Vectorized<float>::loadu(in_out_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                    auto tmp1 = static_cast<float>(0.0);
                    auto tmp2 = at::vec::Vectorized<float>(tmp1);
                    auto tmp3 = at::vec::VecMask<float,1>(tmp0 <= tmp2);
                    auto tmp5 = tmp0 * tmp4;
                    auto tmp6 = decltype(tmp5)::blendv(tmp0, tmp5, tmp3.template cast<float,1>());
                    tmp6.store(in_out_ptr0 + static_cast<int64_t>(x0));
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (1, 1, 100, 100), (10000, 10000, 100, 1))
    # Topologically Sorted Source Nodes: [x], Original ATen: [aten.rrelu_with_noise_functional]
    buf0 = torch.ops.aten.uniform.default(arg0_1, 0.125, 0.3333333333333333)
    buf1 = buf0
    assert_size_stride(buf1, (1, 1, 100, 100), (10000, 10000, 100, 1))
    # buffer buf1 is assumed to be not aligned
    del buf0
    buf2 = buf1; del buf1  # reuse
    cpp_fused_rrelu_with_noise_functional_0(buf2, arg0_1)
    del arg0_1
    return (buf2, )
```

**Comment by shaoyuyoung** [comment]
Got it, thanks!
Is this a bug or an expected behavior?
It seems that I should change the title?

**Comment by CaoE** [comment]
@shaoyuyoung I think this is a expected behavior.

**Comment by shaoyuyoung** [comment]
thanks!
If there is indeed no fix plan, please feel free to close this issue at any time. :)

**Comment by CaoE** [comment]
Please feel free to reopen this issue :)

**Comment by ngimel** [comment]
Even though inductor is expected to generate different values, `config.fallback_random=True` should cause inductor to call eager operation for this, reopening to get it fixed.

**Comment by Yuxingwang-intel** [comment]
@ngimel The following comment in the code explains that the config.fallback_random=True dose not force certain RNG ops to fall back to eager mode. The rrelu and uniform are explicitly registered for decomposition. Once decomposed, these ops are prioritized for functionalization. Therefore config.fallback_random=True has no effect on ops that are already decomposed. This is not a bug but a design tradeoff.
https://github.com/pytorch/pytorch/blob/4c55dc50355d5e923642c59ad2a23d6ad54711e7/torch/_decomp/decompositions_for_rng.py#L221-L242

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260518+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs failed with max_diff=0.8208 between eager and compiled outputs of torch.nn.RReLU() on CPU. allclose checks at rtol/atol=1e-3, 1e-5, 1e-7 all returned False. config.fallback_random=True was set in the repro.

<details>
<summary>Error output</summary>

```
backend: inductor
device: cpu
eager dtype: torch.float32, shape: (1, 1, 100, 100)
max_diff: 0.8208001255989075
allclose(rtol=0.001, atol=0.001): False
allclose(rtol=1e-05, atol=1e-05): False
allclose(rtol=1e-07, atol=1e-07): False

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.rrelu = torch.nn.RReLU()

    def forward(self, x):
        return self.rrelu(x)


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present.

    Issue #147255: torch.nn.RReLU() on CPU produces large numerical mismatch
    between eager and inductor's CPP backend even with
    config.fallback_random=True. The reported max diff is ~0.82.
    """
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    device = "cpu"

    model = Model().to(device)

    torch.manual_seed(0)
    x = torch.randn(1, 1, 100, 100, device=device)

    torch.manual_seed(0)
    eager_out = model(x)

    torch.manual_seed(0)
    compiled_model = torch.compile(model, backend=backend)
    compiled_out = compiled_model(x)

    return torch.allclose(eager_out, compiled_out, rtol=1e-3, atol=1e-3)


if __name__ == "__main__":
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    device = "cpu"

    torch._dynamo.reset()
    model = Model().to(device)

    torch.manual_seed(0)
    x = torch.randn(1, 1, 100, 100, device=device)

    torch.manual_seed(0)
    eager_out = model(x)

    torch.manual_seed(0)
    compiled_out = torch.compile(model, backend=backend)(x)

    max_diff = (eager_out - compiled_out).abs().max().item()
    print(f"backend: {backend}")
    print(f"device: {device}")
    print(f"eager dtype: {eager_out.dtype}, shape: {tuple(eager_out.shape)}")
    print(f"max_diff: {max_diff}")
    for tol in (1e-3, 1e-5, 1e-7):
        print(f"allclose(rtol={tol}, atol={tol}): {torch.allclose(eager_out, compiled_out, tol, tol)}")

    try:
        ok = test_fn()
    except Exception:
        import traceback; traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260518+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260517+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260518+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
